src/ptvsd/adapter/server.py

```diff
@@ -45,12 +161,32 @@ class Capabilities(components.Capabilities):
             "supportedChecksumAlgorithms": [],
         }
 
-    def __init__(self, session, stream):
-        super(Server, self).__init__(session, stream)
+    def __init__(self, session, connection):
+        with _lock:
```

Comment:
The only place I could see creating the server is in `attach_to_session`:

```
with _lock:
    ...
    self.server = Server(session, self)
```

Maybe the `Server` could actually be pseudo-private (i.e.: `_Server`) to notify that it's constructor should not be used anywhere else and then the server constructor wouldn't need to acquire the lock and do `connection.server = self` since this is already taken care of (unless this really needs to be set before the remainder of the constructor -- if that's the case, it'd be nice to add some comment regarding it).

Reply:
The idea was to sync with `ServerConnection.disconnect()`. But looking at it again, it's not doing that correctly, because `Server.disconnect()` can end up getting invoked before there's enough object state present for it to execute correctly.  I was trying to avoid holding `_lock` for the entirety of the constructor here, but it seems unavoidable. And then, yeah, it makes more sense for `attach_to_session()` to take care of it all.